docs/design-documents/features/storage/TDBStore/TDBStore_design.md

```diff
@@ -0,0 +1,497 @@
+- Release `_mutex`
+
+**iterator_next function**
+
+Header:  
+`virtual int iterator_next(iterator_t it, char *key, size_t key_size);`
+
+Pseudo code:
+- Take `_mutex`  
+- While `ram_table_ind` field in iterator smaller than `_num_keys`
+	- Read key pointed to by ram table in `ram_table_ind` into a local variable
+	- If name matches prefix
+		- Advance `ram_table_ind` field in iterator
+		- Copy name to `key` and return OK
+	- Advance `ram_table_ind` field in iterator 
+- Return "not found" error
+- Release `_mutex`
+
+**iterator_close function**
+
+Header:  
+`virtual int iterator_close(iterator_t it);`
+
+Pseudo code:
+- Release `prefix` field in iterator and structure allocated at `it`
+
+
+### Reserved space
+
+**reserved_space_set function**
+
+Header:  
+`virtual int reserved_space_set(void *data);`
+
+Pseudo code:  
+- Check if reserved space is not empty, if it is, return a "reserved space programmed error"
+- Copy `data` contents to reserved space location 
+
+**reserved_space_get function**
+
+Header:  
+`virtual int reserved_space_get(void *data);`
+
+Pseudo code:  
+- Copy contents from reserved space location `data`  
+
+
+# Usage scenarios and examples
+
+### Standard usage of the class
+
+Following example code shows standard usage of the TDBStore class 
+
+**Standard usage example**
+
+```C++
+// Underlying block device. Here, SPI Flash is fully used.
+// One can use SlicingBlockDevice if we want a partition.
+SPIFBlockDevice bd(PTE2, PTE4, PTE1, PTE5);
+
+// Instantiate tdbstore with our block device and 64 as max number of keys
+TDBStore tdbstore(64, &bd);
+
+int res;
+
+// Initialize tdbstore
+res = tdbstore.init();
+
+// Add "Key1"
+const char *val1 = "Value of key 1";
+const char *val2 = "Updated value of key 1";
+res = tdbstore.set("Key1", val1, sizeof(val1), 0);
+// Update value of "Key1"
+res = tdbstore.set("Key1", val2, sizeof(val2), 0);
+
+uint_8 value[32];
+size_t actual_size;
+// Get value of "Key1". Value should return the updated value.
+res = tdbstore.get("Key1", value, sizeof(value), &actual_size);
+
+// Remove "Key1"
+res = tdbstore.remove("Key1");
+
+// Incremental write, if need to generate large data with a small buffer
+const int data_size = 1024;
+char buf[8];
+
+KVSTore::set_handle_t handle;
+res = tdbstore.set_start(&handle, "Key2", data_size, 0);
+for (int i = 0; i < data_size / sizeof(buf); i++) {
+	memset(buf, i, sizeof(buf));
+	res = tdbstore.set_add_data(handle, buf, sizeof(buf));
+}
+res = tdbstore.set_finalize(handle);
+
+// Iterate over all keys starting with "Key"
+res = 0;
+KVSTore::iterator_t it;
+tdbstore.iterator_open(&it, "Key*");
+char key[KVSTore::KV_MAX_KEY_LENGTH];
+while (!res) {
+    res = tdbstore.iterator_next(&it, key, sizeof(key));
+}
+res = tdbstore.iterator_close(&it);
+
+// Deinitialize TDBStore
+res = tdbstore.deinit();
+```
+# Other information
+
+### Open issues
+
+- Need to figure a way to prevent mutex abuse in incremental set APIs. 
```

Comment:
I think this is just a risk of using the advanced increment set APIs. May be worth a big warning note in the doxygen on the functions.

Also we should require that all KVStore implementations MUST hold a mutex if they're thread safe. Even the FileSystemStore where it's not strictly necessary. Otherwise KVStore <-> FileSystemStore isn't portable.

May be worth noting in the KVStore design doc.